clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,63 +348,560 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+                    //  to the subscription state at this point.
+                }
+            }
+        });
+
+        return true;
+    }
+
+    /**
+     *  Visible for testing.
+     */
+    void markReconciliationInProgress() {
+        reconciliationInProgress = true;
+        memberIdOnReconciliationStart = memberId;
+    }
+
+    /**
+     *  Visible for testing.
+     */
+    void markReconciliationCompleted() {
+        reconciliationInProgress = false;
+    }
+
+    /**
+     * Build set of TopicPartition (topic name and partition id) from the target assignment
+     * received from the broker (topic IDs and list of partitions).
+     *
+     * <p>
+     * This will:
      *
-     * @throws IllegalStateException If a target assignment already exists.
+     * <ol type="1">
+     *     <li>Try to find topic names in the metadata cache</li>
+     *     <li>For topics not found in metadata, try to find names in the local topic names cache
+     *     (contains topic id and names currently assigned and resolved)</li>
+     *     <li>If there are topics that are not in metadata cache or in the local cached
+     *     of topic names assigned to this member, request a metadata update, and continue
+     *     resolving names as the cache is updated.
+     *     </li>
+     * </ol>
      */
-    private void setTargetAssignment(ConsumerGroupHeartbeatResponseData.Assignment newTargetAssignment) {
-        if (!targetAssignment.isPresent()) {
-            log.info("Member {} accepted new target assignment {} to reconcile", memberId, newTargetAssignment);
-            targetAssignment = Optional.of(newTargetAssignment);
+    private void resolveMetadataForUnresolvedAssignment() {
+
+        // Try to resolve topic names from metadata cache or subscription cache, and move
+        // assignments from the "unresolved" collection, to the "readyToReconcile" one.
+        Iterator<Map.Entry<Uuid, List<Integer>>> it = assignmentUnresolved.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<Uuid, List<Integer>> e = it.next();
+            Uuid topicId = e.getKey();
+            List<Integer> topicPartitions = e.getValue();
+
+            Optional<String> nameFromMetadata = findTopicNameInGlobalOrLocalCache(topicId);
+            if (nameFromMetadata.isPresent()) {
+                // Name resolved, so assignment is ready for reconciliation.
+                assignmentReadyToReconcile.addAll(buildAssignedPartitionsWithTopicName(topicPartitions,
+                        nameFromMetadata.get()));
+                it.remove();
+            }
+        }
+
+        if (!assignmentUnresolved.isEmpty()) {
+            log.debug("Topic Ids {} received in target assignment were not found in metadata and " +
+                    "are not currently assigned. Requesting a metadata update now to resolve " +
+                    "topic names.", assignmentUnresolved.keySet());
+            metadata.requestUpdate(true);
```

Comment:
I am not sure to understand how the metadata cache knows which new topic ids it should resolve. Or does the consumer request metadata for ALL topics in the cluster? Looking at the code, it is may be what it does.

Reply:
Yes, that's what it does, get metadata for all topics [here](https://github.com/apache/kafka/blob/22f7ffe5e1623d279096b45ab475768eeb05eee1/clients/src/main/java/org/apache/kafka/clients/Metadata.java#L719). It seems that there was an intention of a partial update [here](https://github.com/apache/kafka/blob/22f7ffe5e1623d279096b45ab475768eeb05eee1/clients/src/main/java/org/apache/kafka/clients/Metadata.java#L700) but not fully implemented, so it effectively ends up getting them all anyways.

Reply:
Let's file a jira to improve this as well. Ideally, assuming that we don't use client side regex anymore, the client should only request the topics that it needs.

Reply:
Makes sense, [KAFKA-15847](https://issues.apache.org/jira/browse/KAFKA-15847)